server/src/main/java/org/elasticsearch/index/translog/EncryptedFileChannel.java

```diff
@@ -0,0 +1,188 @@
+        return this.channel.write(ByteBuffer.wrap(tmp));
+    }
+
+    @Override
+    public int write(ByteBuffer src, long position) throws IOException {
+        byte[] tmp = this.cipher.encrypt(src, position);
+        return this.channel.write(ByteBuffer.wrap(tmp), position);
+    }
+
+    @Override
+    public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long position() throws IOException {
+        return this.channel.position();
+    }
+
+    @Override
+    public FileChannel position(long newPosition) throws IOException {
+        this.channel.position(newPosition);
+        return this;
+    }
+
+    @Override
+    public long size() throws IOException {
+        return this.channel.size();
+    }
+
+    @Override
+    public FileChannel truncate(long size) throws IOException {
+        this.channel.truncate(size);
+        return this;
+    }
+
+    @Override
+    public void force(boolean metaData) throws IOException {
+        this.channel.force(metaData);
+    }
+
+    @Override
+    public long transferTo(long position, long count, WritableByteChannel target) throws IOException {
+        long tmpPosition = position;
+        long tmpCount = count;
+        long transferCount = 0;
+        ByteBuffer bb = ByteBuffer.allocate(MAX_TRANSFER_BUFFER_SIZE);
+
+        while (tmpCount > 0){
+            int read = this.read(bb, tmpPosition);
+            if (read <= 0) {
+                break;
+            }
+
+            int dataSize = tmpCount < read ? (int) tmpCount : read;
+            tmpCount -= dataSize;
+
+            bb.flip();
+            byte[] data = new byte[dataSize];
+            bb.get(data, 0, dataSize);
+
+            int write = target.write(ByteBuffer.wrap(data));
+            tmpPosition += write;
+            transferCount += write;
+
+            if (bb.remaining() > 0) {
+                break;
+            }
+            bb.flip();
+        }
+
+        return transferCount;
+    }
+
+    @Override
+    public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException {
+        long tmpPosition = position;
+        long tmpCount = count;
+        long transferCount = 0;
+        ByteBuffer bb = ByteBuffer.allocate(MAX_TRANSFER_BUFFER_SIZE);
+
+        while (tmpCount > 0){
+            int read = src.read(bb);
+            if (read <= 0) {
+                break;
+            }
+
+            int dataSize = tmpCount < read ? (int) tmpCount : read;
+            tmpCount -= dataSize;
+
+            bb.flip();
+            byte[] data = new byte[dataSize];
+            bb.get(data, 0, dataSize);
+            bb.flip();
+
+            int write = this.write(ByteBuffer.wrap(data), tmpPosition);
+            tmpPosition += write;
+            transferCount += write;
+        }
+
+        return transferCount;
```

Comment:
Same situation as in `transferTo`
-  `this.channel`'s position should not be modified
- we should never read more than `count` bytes from `src` (this implementation attempts to read `8192`)

Reply:
Same goes for `FileChannel.write(ByteBuffer src, long position)`.

Reply:
But `src.read(bb)` would move the `src` position by `MAX_TRANSFER_BUFFER_SIZE`, right? (in case that the `src` channel implementation supports `position` of course).
And the expected behaviour would be to move it by `count` at most.

Reply:
Yes, you are right, I didn't realize that. I'll fix it.